caikit/core/data_model/data_backends/dict_backend.py

```diff
@@ -103,4 +103,21 @@ def get_attribute(self, data_model_class: Type[DataBase], name: str) -> Any:
                 for entry in raw_value
             ]
 
+        # If the target attribure is a oneof name, look for the oneof name in the data dict first.
+        # If it exists, infer the right field name based on the type of the value and return the
+        # field name and the value. If oneof name isn't in the data dict, check for field name
+        # instead and return the right value
+        if name in data_model_class._fields_oneofs_map:
+            if name in self._data_dict:
+                val = self._data_dict.get(name)
+                which_oneof = data_model_class._infer_which_oneof(name, val)
+                return data_model_class.OneofFieldVal(val=val, which_oneof=which_oneof)
+
+            oneof_fields = data_model_class._fields_oneofs_map.get(name)
+            for field in oneof_fields:
+                if field in self._data_dict:
+                    return data_model_class.OneofFieldVal(
+                        val=self._data_dict.get(field), which_oneof=field
+                    )
+
```

Comment:
I think we should also include the following

```py
if name in data_model_class._fields_to_oneof:
    if name in self._data_dict:
        return data_model_class.OneofFieldVal(val=self._data_dict[name], which_oneof=name)
    oneof_name = data_model_class._fields_to_oneof[name]
    if oneof_name in self._data_dict and data_model_class._infer_which_oneof(oneof_name, self._data_dict[val]) == name:
        return data_model_class.OneofFieldVal(val=self._data_dict[name], which_oneof=name)
    return None
```

Reply:
(of course with `fmt` applied 😉)

Reply:
I _think_ this works. I think it would be good to add a couple more test cases to get pedantic:

```py
msg3 = Foo.from_backend(DictBackend({"fooint": 1234}))
assert msg2.to_dict() == msg3.to_dict()
assert msg2.foo == msg3.foo
assert msg2.fooint == msg3.fooint
assert msg2.foostr == msg3.foostr == None
```